src/java.base/share/classes/jdk/internal/foreign/ArenaImpl.java

```diff
@@ -52,13 +49,14 @@ public void close() {
 
     public NativeMemorySegmentImpl allocateNoInit(long byteSize, long byteAlignment) {
         Utils.checkAllocationSizeAndAlign(byteSize, byteAlignment);
-        return SegmentFactories.allocateSegment(byteSize, byteAlignment, session, shouldReserveMemory);
+        long address = SegmentFactories.allocateNative(byteSize, byteAlignment, session, shouldReserveMemory, false);
+        return new NativeMemorySegmentImpl(address, byteSize, false, session);
```

Comment:
Could you move this constructor call to `allocateNative`? All segment construction calls are currently in `SegmentFactories` as a measure to avoid bootstrap cycles, which we had problems with in the past.

Reply:
I want to put it here so that even if `SegmentFactories::allocateNative` is not inlined, this small method will be more likely to be inlined and the segment can be non-escape, then. I have added a comment regarding class initialization cycles.

Reply:
I understand, but can't we then move this method to `SegmentFactories` - e.g. maybe have a private `allocateNativeInternal` and an `allocateNative` ?

Reply:
That's reasonable. I have done so.